pkgs/build-support/fetchgit/nix-prefetch-git

```diff
@@ -148,7 +148,17 @@ checkout_hash(){
         hash=$(hash_from_ref "$ref")
     fi
 
-    clean_git fetch -t ${builder:+--progress} origin || return 1
+    if [[ -n "$deepClone" ]]; then
+        clean_git fetch -t ${builder:+--progress} origin || return 1
+        # Try to shallow fetch specific hash if it doesn't exist in any ref.
+        # For example, if a ref was force-pushed but the server still exposes the old hash.
+        clean_git fetch ${builder:+--progress} --depth 1 origin "$hash" 2> /dev/null
+    else
+        # Try to shallow fetch specific hash if server supports uploadpack.allowReachableSha1InWant.
+        # Fallback to full fetch on failure.
+        clean_git fetch ${builder:+--progress} --depth 1 origin "$hash" 2> /dev/null ||
+        clean_git fetch -t ${builder:+--progress} origin || return 1
+    fi
```

Comment:
Assuming the shallow fetch succeeds, won't the derivation's output hash change from that of a full fetch? So, if someone uses `leaveDotGit = true` they'll also have to change their sha256 field?

Reply:
`make_deterministic_repo` removes all remote branches & tags not reachable from head, so a shallow fetch will result in the same output. It looks like there's a problem with the `deepClone` though in `update-nix-fetchgit`'s tests:
```
FAIL (0.77s)
      Files 'tests/test_dotgit.expected.nix' and 'tests/test_dotgit.out.nix' differ
```
```diff
--- test_dotgit.expected.nix
+++ test_dotgit.out.nix
@@ -1,25 +1,25 @@
 { pkgs ? import <nixpkgs> { } }:
 
 with pkgs;
 
 {
   src = fetchgit {
     url = "/tmp/nix-update-fetchgit-test/repo1";
     rev = "1c60ae07b5740aab02e32b4f64600f002112e6fd";
     sha256 = "008xwkjfsv3rj50x9fqj0vvggl1zwrbqh7bkvnga94krmij63hib";
   };
 
   srcDotGit = fetchgit {
     url = "/tmp/nix-update-fetchgit-test/repo1";
     rev = "1c60ae07b5740aab02e32b4f64600f002112e6fd";
     sha256 = "1ywf4ysn3fcr4dw24idz48i36ghvxbvi72icglp4nqhdj2s6bzj1";
     leaveDotGit = true;
   };
 
   srcDeep = fetchgit {
     url = "/tmp/nix-update-fetchgit-test/repo1";
     rev = "1c60ae07b5740aab02e32b4f64600f002112e6fd";
-    sha256 = "0xnazgn7p7xspda2wsfk1v1xrw9z5vvban0nhwd7xd3f70223gzz";
+    sha256 = "0fwgql4qxiyaz3knib422hbq29cjcqdas87xk7qk3qcj20z7hnby";
     deepClone = true;
   };
 }
```

Strangely, removing the shallow fetch for deep cloning fixes the test failure :confused:. I assumed that if the commit already existed in a ref, it would just be a no-op. I'm looking into it now.